File: NestedSat.m
%% System Ax + Bu
A     = [0 1 0 0; 0 0 1 0; 0 0 0 1; 0 0 0 0];
B     = [0; 0; 0; 9.8]; 

%Values from paper
p     = 2*[0.5,1,2,3]; %Poles
% b     = [0.625,1.25,2.5,5]; %Sat values
b     = 100*[0.625,1.25,2.5,5]; %Sat values

%Transform matrix
Tx = [p(1)*p(2)*p(3)*p(4), (p(1)*p(3) + p(2)*p(3) + p(1)*p(2))*p(4), (p(1) + p(2) + p(3))*p(4),     p(4);
            0                           p(1)*p(2)*p(3)                  (p(1) + p(2))*p(3),         p(3);
            0                                   0                               p(1)*p(2)           p(2);
            0                                   0                                   0               p(1)]; 


dt    = 0.001;  %Time step
t     = 50;     %Sim time
s     = 0:dt:t; %Time

x = zeros(size(A,1),length(s)); %State vector
q(:,1) = [0; 0; 0; 0];    %Initial condotions
e = zeros(size(A,1),length(s)); %Error vector
r = zeros(size(A,1),length(s)); %Reference vector
zx = zeros(size(A,1),length(s)); %Transform vector
u = zeros(1,length(s));

%% Solver and control
for i = 1 : length(s)
    r(:,i) = [4 + sin(0.5*(i-1)*dt); 0.5*cos(0.5*(i-1)*dt); -0.25*sin(0.5*(i-1)*dt); -0.125*cos(0.5*(i-1)*dt)];%Reference
    e(:,i) = x(:,i) - r(:,i); %Error
    xpppp  = 0.0625*sin(0.5*(i-1)*dt);

    zx(:,i) = Tx*e(:,i);

    u(i) = xpppp-(sat(zx(4,i)+sat(zx(3,i)+sat(zx(2,i)+sat(zx(1,i),b(1)),b(2)),b(3)),b(4)))/9.8; %Nested saturations

    x(:,i+1) = x(:,i) + dt*(A*x(:,i) + B*u(i)- [0;0;0;xpppp]); %Forward Euler
end

%% Plots
%Position plot
figure(1)
grid on; hold on;
plot(s,x(1,1:end-1), 'LineWidth', 2)
plot(s,r(1,:), 'LineWidth', 2);
xlabel({'Time(s)'}, 'interpreter', 'latex', 'fontsize', 14)
ylabel({'Position(m)'}, 'interpreter', 'latex', 'fontsize', 14)
legend({'$x_{1}(t)$','$x_{1ref}(t)$'}, 'interpreter', 'latex', 'location', 'best', 'fontsize', 14);


%Errors (e1->x, e2->xp, e3->xpp, e4->xppp)
figure(2)
grid on; hold on;
plot(s,e, 'LineWidth', 2);
xlabel({'Time'}, 'interpreter', 'latex', 'fontsize', 14)
ylabel({'Errors'}, 'interpreter', 'latex', 'fontsize', 14)
legend({'$e_{1}(t)$','$e_{2}(t)$','$e_{3}(t)$','$e_{4}(t)$'}, 'interpreter', 'latex', 'location', 'best', 'fontsize', 14);

%Control input plot
figure(3)
grid on; hold on;
plot(s,u, 'LineWidth', 2)
xlabel({'Time'}, 'interpreter', 'latex', 'fontsize', 14)
ylabel({'Control input'}, 'interpreter', 'latex', 'fontsize', 14)


%% Auxiliar function --- Sat(.) ---
function res = sat(valor,limite)
    if valor > limite
        res = limite;
    elseif valor < -limite
        res = -limite;
    else
        res = valor;
    end
end

File: NestedSat.py
"""
This script is based on the article
Nested saturation with guaranteed real poles
https://ieeexplore.ieee.org/document/1239062
"""

import numpy as np
import matplotlib.pyplot as plt

# --- Saturation function ---
def sat(value, limit):
    return np.clip(value, -limit, limit)

# --- System matrices ---
A = np.array([[0, 1, 0, 0],
              [0, 0, 1, 0],
              [0, 0, 0, 1],
              [0, 0, 0, 0]])

B = np.array([0, 0, 0, 9.8])

# --- Parameters from paper ---
p = 2 * np.array([0.5, 1, 2, 3])     # poles
b = 100 * np.array([0.625, 1.25, 2.5, 5])  # saturation limits

# --- Transformation matrix Tx ---
Tx = np.array([
    [p[0]*p[1]*p[2]*p[3],
     (p[0]*p[2] + p[1]*p[2] + p[0]*p[1]) * p[3],
     (p[0] + p[1] + p[2]) * p[3],
     p[3]],

    [0,
     p[0]*p[1]*p[2],
     (p[0] + p[1]) * p[2],
     p[2]],

    [0, 0,
     p[0]*p[1],
     p[1]],

    [0, 0, 0,
     p[0]]
])

# --- Simulation parameters ---
dt = 0.001
t_final = 50
s = np.arange(0, t_final + dt, dt)

# --- State variables ---
x = np.zeros((4, len(s)))
e = np.zeros((4, len(s)))
r = np.zeros((4, len(s)))
zx = np.zeros((4, len(s)))
u = np.zeros(len(s))

# Initial condition (already zero, but explicit)
x[:, 0] = np.array([0, 0, 0, 0])

# --- Simulation loop ---
for i in range(len(s) - 2):
    t = s[i]

    # Reference
    r[:, i] = np.array([
        4 + np.sin(0.5 * t),
        0.5 * np.cos(0.5 * t),
        -0.25 * np.sin(0.5 * t),
        -0.125 * np.cos(0.5 * t)
    ])

    # Error
    e[:, i] = x[:, i] - r[:, i]

    # 4th derivative reference
    xpppp = 0.0625 * np.sin(0.5 * t)

    # Transformation
    zx[:, i] = Tx @ e[:, i]

    # --- Nested saturation ---
    inner1 = sat(zx[0, i], b[0])
    inner2 = sat(zx[1, i] + inner1, b[1])
    inner3 = sat(zx[2, i] + inner2, b[2])
    inner4 = sat(zx[3, i] + inner3, b[3])

    u[i] = xpppp - inner4 / 9.8

    # System dynamics (Euler integration)
    x_dot = A @ x[:, i] + B * u[i] - np.array([0, 0, 0, xpppp])
    x[:, i + 1] = x[:, i] + dt * x_dot

# --- Plot 1: Position ---
plt.figure()
plt.plot(s, x[0, :], label='x1')
plt.plot(s, r[0, :], label='x1_ref')
plt.xlabel("Time (s)")
plt.ylabel("Position (m)")
plt.legend()
plt.grid()

# --- Plot 2: Errors ---
plt.figure()
for i in range(4):
    plt.plot(s, e[i, :], label=f'e{i+1}')
plt.xlabel("Time")
plt.ylabel("Errors")
plt.legend()
plt.grid()

# --- Plot 3: Control input ---
plt.figure()
plt.plot(s, u, label='u')
plt.xlabel("Time")
plt.ylabel("Control input")
plt.grid()

plt.show()